core/src/main/java/com/scalar/db/util/ActiveExpiringMap.java

```diff
@@ -91,14 +105,98 @@ public V remove(K key) {
     return prev.get();
   }
 
+  public Set<K> keySet() {
+    return map.keySet();
+  }
+
+  public Collection<V> values() {
+    Collection<ValueHolder<V>> values = map.values();
+    return new AbstractCollection<V>() {
+      @Override
+      public Iterator<V> iterator() {
+        return new Iterator<V>() {
+          private final Iterator<ValueHolder<V>> valueIterator = values.iterator();
+
+          @Override
+          public boolean hasNext() {
+            return valueIterator.hasNext();
+          }
+
+          @Override
+          public V next() {
+            ValueHolder<V> next = valueIterator.next();
+            return next.get();
+          }
+
+          @Override
+          public void remove() {
+            valueIterator.remove();
+          }
+        };
+      }
+
+      @Override
+      public int size() {
+        return values.size();
+      }
+    };
+  }
+
+  public Set<Map.Entry<K, V>> entrySet() {
+    return new AbstractSet<Entry<K, V>>() {
+      private final Set<Entry<K, ValueHolder<V>>> entries = map.entrySet();
+
+      @Override
+      public Iterator<Entry<K, V>> iterator() {
+        return new Iterator<Entry<K, V>>() {
+          private final Iterator<Entry<K, ValueHolder<V>>> entryIterator = entries.iterator();
+
+          @Override
+          public boolean hasNext() {
+            return entryIterator.hasNext();
+          }
+
+          @SuppressFBWarnings("SE_BAD_FIELD")
+          @Override
+          public Entry<K, V> next() {
+            Entry<K, ValueHolder<V>> next = entryIterator.next();
+            return new AbstractMap.SimpleEntry<K, V>(next.getKey(), next.getValue().get()) {
+              @Override
```

Comment:
Is this override needed?

Reply:
You mean whether or not we really need to override the `setValue()` method and make it throw `UnsupportedOperationException`, right?

We expect that `setValue()` replaces the actual value of the map's entry, but I don't think we need it now, so I override it.

Do you think we need to implement it?

Reply:
> but I don't think we need it now, so I override it.

I got your point. Sounds good!

The class doesn't seem to guarantee immutability, so I thought we also can allow users to modify values. But I don't have any strong opinion on it.

Reply:
>  I thought we also can allow users to modify values
Ah, unlike java.util.HashMap which exposes internal HashMap.Node via iterator, ActiveExpiringMap creates an entry copying the original key and value. So, we need to do something if we want ActiveExpiringMap to behave like HashMap. I agree with your original idea now.